### PR TITLE
Fixed Change Feed Metadata timestamp de-serialization

### PR DESCRIPTION
**Pull request by kushagraThapar**
`ChangeFeedMetaData` has Instant type for `conflictResolutionTimestamp` which with certain values of timestamps (ending with 0), Instant class deserializes them with nanoseconds in exponential format. This can break down the parsing of these fields for customers. 

One such example is this: 

Note the "crts" timestamp - 
Before the fix: 

FFCF changes are :` {"current":{"id":"1946c04a-070b-48c0-8e36-517c8d2f92ed","mypk":"mypk-1","prop":"Johnson","_rid":"NopBALG34lcBAAAAAAAAAA==","_self":"dbs/NopBAA==/colls/NopBALG34lc=/docs/NopBALG34lcBAAAAAAAAAA==/","_etag":"\"00000000-0000-0000-b857-fd9e73fb01d9\"","_attachments":"attachments/","_ts":1689561604},"metadata":{"crts":1.6895616E+9,"lsn":176,"operationType":"create","previousImageLSN":0,"timeToLiveExpired":false}}`

* Fixed Change Feed Metadata timestamp de-serialization

After the fix: 

FFCF changes are :` {"current":{"id":"1946c04a-070b-48c0-8e36-517c8d2f92ed","mypk":"mypk-1","prop":"Johnson","_rid":"NopBALG34lcBAAAAAAAAAA==","_self":"dbs/NopBAA==/colls/NopBALG34lc=/docs/NopBALG34lcBAAAAAAAAAA==/","_etag":"\"00000000-0000-0000-b857-fd9e73fb01d9\"","_attachments":"attachments/","_ts":1689561604},"metadata":{"crts":1689561600,"lsn":176,"operationType":"create","previousImageLSN":0,"timeToLiveExpired":false}}`

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ChangeFeedProcessorItemSerializerTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        Assertions.assertThat(jsonNode.get("metadata").get("crts").asText()).isEqualTo("1689561600");
+    }
+
+    @Test(groups = { "unit" })
+    public void testReplaceChangeFeedProcessorItemDeSerializer() throws JsonProcessingException {
+        String json = "{\"current\":{\"id\":\"1946c04a-070b-48c0-8e36-517c8d2f92ed\",\"mypk\":\"mypk-1\",\"prop\":\"Gates\",\"_rid\":\"NopBALG34lcBAAAAAAAAAA==\",\"_self\":\"dbs/NopBAA==/colls/NopBALG34lc=/docs/NopBALG34lcBAAAAAAAAAA==/\",\"_etag\":\"\\\"00000000-0000-0000-b857-fda256f201d9\\\"\",\"_attachments\":\"attachments/\",\"_ts\":1689561604},\"metadata\":{\"lsn\":178,\"crts\":1689561600,\"operationType\":\"replace\",\"previousImageLSN\":176}}";
+        ChangeFeedProcessorItem changeFeedProcessorItem = simpleObjectMapper.readValue(json, ChangeFeedProcessorItem.class);
+        JsonNode jsonNode = changeFeedProcessorItem.toJsonNode();
+        Assertions.assertThat(jsonNode.get("metadata").get("crts").asText()).isEqualTo("1689561600");
+    }
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ChangeFeedProcessorItemSerializerTest.java:42]: Non-blocking - could we have a test to see if this doesn't introduce regressions on `crts` which doesn't end with 0?

**Comment by jeet1995** [comment]
This would also need to have a CHANGELOG.md update for 4.48.0, correct?

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ChangeFeedProcessorItemSerializerTest.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/ChangeFeedProcessorItemSerializerTest.java:42]: Added more in next iteration.

**Comment by kushagraThapar** [comment]
> This would also need to have a CHANGELOG.md update for 4.48.0, correct?

Added in next iteration.

**Review by milismsft — APPROVED** [review]

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
Sorry - just saw that the same problem can occur for _ts as well. Please use the same model for _ts (and ass test coverage there as well)?

**Comment by kushagraThapar** [comment]
> Sorry - just saw that the same problem can occur for _ts as well. Please use the same model for _ts (and ass test coverage there as well)?

_ts is not exposed as `Instant` anywhere in the change feed, so this problem cannot happen with _ts

**Review by kirankumarkolli — APPROVED** [review]

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java [hunk 2]
```diff
@@ -55,8 +58,14 @@ public JsonNode getPrevious() {
      * @return change feed metadata.
      */
     @Beta(value = Beta.SinceVersion.V4_37_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
+    @JsonIgnore
     public ChangeFeedMetaData getChangeFeedMetaData() {
-        return changeFeedMetaData;
+        synchronized (monitor) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java:63]: We don't need this synchronization for the getter. We can just return a new instance. Worst case we create a couple of extra instances, but that's okay.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java [hunk 3]
```diff
@@ -17,6 +19,7 @@
  */
 @Beta(value = Beta.SinceVersion.V4_37_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
 public final class ChangeFeedProcessorItem {
+    private final static ObjectMapper mapper = Utils.getSimpleObjectMapper();
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java:22]: We should not store this, as it just returns a reference, so we can call it as many times. 
Lets not make a new reference.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java [hunk 4]
```diff
@@ -69,7 +89,18 @@ public ChangeFeedMetaData getChangeFeedMetaData() {
      */
     @Beta(value = Beta.SinceVersion.V4_37_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
     public JsonNode toJsonNode() {
-        return Utils.getSimpleObjectMapper().convertValue(this, JsonNode.class);
+
+        if (this.changeFeedProcessorItemAsJsonNode == null) {
+            this.changeFeedProcessorItemAsJsonNode = Utils.getSimpleObjectMapper().convertValue(this, JsonNode.class);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java:94]: We don't need to convert, we can create an ObjectNode and then set current, previous and metadata.

Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java]: Fixed in next iteration.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java (hunk 2 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java]: Fixed in next iteration.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java [hunk 5]
```diff
@@ -55,8 +60,18 @@ public JsonNode getPrevious() {
      * @return change feed metadata.
      */
     @Beta(value = Beta.SinceVersion.V4_37_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
+    @JsonIgnore
     public ChangeFeedMetaData getChangeFeedMetaData() {
-        return changeFeedMetaData;
+
+        if (this.changeFeedProcessorItemAsJsonNode == null) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java:66]: why we need to construct changeFeedProcessorItemAsJsonNode? does not seem to be used in this method

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java [hunk 6]
```diff
@@ -55,8 +60,18 @@ public JsonNode getPrevious() {
...
+            this.changeFeedProcessorItemAsJsonNode = constructChangeFeedProcessorItemAsJsonNode();
+        }
+
+        if (this.changeFeedMetaDataInternal == null) {
+            this.changeFeedMetaDataInternal = Utils.getSimpleObjectMapper().convertValue(this.changeFeedMetaData, ChangeFeedMetaData.class);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java:71]: do we need to validate changeFeedMetaData is not null?

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java (hunk 6 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java:67]: `convertValue()` will return `null` with no exceptions. So, no additional check needed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java [hunk 7]
```diff
@@ -69,15 +84,48 @@ public ChangeFeedMetaData getChangeFeedMetaData() {
      */
     @Beta(value = Beta.SinceVersion.V4_37_0, warningText = Beta.PREVIEW_SUBJECT_TO_CHANGE_WARNING)
     public JsonNode toJsonNode() {
-        return Utils.getSimpleObjectMapper().convertValue(this, JsonNode.class);
+
+        if (this.changeFeedProcessorItemAsJsonNode == null) {
+            this.changeFeedProcessorItemAsJsonNode = constructChangeFeedProcessorItemAsJsonNode();
+        }
+
+        if (this.changeFeedMetaDataInternal == null) {
```
[the remark this answers is not in the document]
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java]: Fixed in next iteration.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java (hunk 5 above)
Comment (jeet1995) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/ChangeFeedProcessorItem.java]: Fixed in next iteration.

**Pull request merged**